test/parallel/test-tls-socket-constructor-alpn-npn-options-parsing.js

```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+// Test that TLSSocket can take arrays of strings for ALPNProtocols and
+// NPNProtocols.
+
+const common = require('../common');
+
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const tls = require('tls');
+
+new tls.TLSSocket(null, {
+  ALPNProtocols: ['http/1.1'],
+  NPNProtocols: ['http/1.1']
+});
+
+if (!process.features.tls_npn)
+  common.skip('node compiled without NPN feature of OpenSSL');
+
+if (!process.features.tls_alpn)
+  common.skip('node compiled without ALPN feature of OpenSSL');
+
+const assert = require('assert');
+const net = require('net');
+const fixtures = require('../common/fixtures');
+
+const key = fixtures.readKey('agent1-key.pem');
+const cert = fixtures.readKey('agent1-cert.pem');
+
+const server = net.createServer(common.mustCall((s) => {
+  const tlsSocket = new tls.TLSSocket(s, {
+    isServer: true,
+    server: server,
+
+    ALPNProtocols: ['a', 'b', 'c'],
+    NPNProtocols: ['d', 'e', 'f'],
+
+    secureContext: tls.createSecureContext({
+      key: key,
+      cert: cert
+    }),
+  });
+
+  tlsSocket.on('secure', common.mustCall((...args) => {
+    // TODO
```

Comment:
@bnoordhuis It seems like this event is the one I should be listening for, but it's undocumented. Am I correct? If so, which object/properties do I check to complete the test?

Reply:
You probably want `'secureConnect'`, `'secure'` is an older, deprecated version.

Once the connection is established, verify the `.alpnProtocol` and `.npnProtocol` properties.

Reply:
`'secureConnect'` is only called when the socket is created using `tls.connect` (and not using the `TLSSocket` constructor directly), so that event will never be emitted. However, the code in `tls.connect` appears to listen for `'secure'` and then do some unrelated synchronous work before emitting `'secureConnect'`.

[1]: https://github.com/nodejs/node/pull/10846